### PR TITLE
[0.9.x] KOGITO-1756 + KOGITO-1779 - Support for Parallel State and fix for SubFlow state support

### PR DESCRIPTION
**Pull request by tsurdilo**

### CODE REVIEW

**Comment by ricardozanini** [comment]
@MarianMacik could you please take a careful look in this one? Many thanks!

**Comment by tsurdilo** [comment]
@MarianMacik I don't think this is the right decision to make. I don't think there was a hard freeze and this code should be merged. Has this decision not to merge made a manager approval?

**Comment by jstastny-cz** [comment]
@tsurdilo it should and is currently being discussed. And yes, there was a hard freeze. Please join us at todays meeting.

**Comment by tsurdilo** [comment]
@lazarotti @krisv Can you guys please help with this -- i think the QE decision just not to merge this is not right thing to do.

**Comment by tsurdilo** [comment]
@jstastny-cz ok well I am glad that this is being discussed because it is not fair for us to work weekends and days off and do all the QE and testing on an off day just to be told hard no for no reason at all.

**Comment by tsurdilo** [comment]
@MarianMacik what changes are requested here? Don't see any. Or is maybe putting the pr in changes requested mode just some sort of blocker?

**Comment by MarianMacik** [comment]
@tsurdilo Well there is no nack on GitHub so what I meant by changes requested was to make this just a blocker fix if the SubFlow issue here is a blocker. But given that none of these issues is a blocker I would merge this just to master. Code Freeze for new features was 3 weeks ago, so I see that there is a reason to not merge it.

**Comment by tsurdilo** [comment]
@MarianMacik well then go and remove the 0.9x Serverless Workflow stuff thats already in. It's much worse to release something 1/2 way than none at all imo, especially for dev preview.

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java [hunk 1]
```diff
@@ -174,6 +175,20 @@ public SubProcessNode callActivity(long id, String name, String calledId, boolea
...
+        VariableScope variableScope = new VariableScope();
+        subProcessNode.addContext(variableScope);
+        subProcessNode.setDefaultContext(variableScope);
+
+        Map<String, String> inputOtuputTypes = new HashMap<>();
+        inputOtuputTypes.put(DEFAULT_WORKFLOW_VAR, JSON_NODE);
+        subProcessNode.setMetaData("BPMN.InputTypes", inputOtuputTypes);
+        subProcessNode.setMetaData("BPMN.OutputTypes", inputOtuputTypes);
+
+        // parent and sub processes have process var "workflowdata"
```
Comment (MarianMacik) [inline jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java:189]: Probably would be better to omit it as if the value of the `DEFAULT_WORKFLOW_VAR` is changed, this comment becomes misleading.

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java [hunk 2]
```diff
@@ -269,23 +284,45 @@ public CompositeContextNode subProcessNode(long id, String name, NodeContainer n
...
     }
 
-    public Split xorSplitNode(long id, String name, NodeContainer nodeContainer) {
+
+    public Split splitNode(long id, String name, int type, NodeContainer nodeContainer) {
+        // 0 = TYPE_UNDEFINED
+        // 1 = TYPE_AND
+        // 2 = TYPE_XOR
+        // 3 = TYPE_OR
+        // 4 = TYPE_XAND
```
Comment (MarianMacik) [inline jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java:298]: Why do we need this here? Can we just put here a link to the documentation or name of the class where this is specified? This quickly can become obsolete once the code is changed.

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java [hunk 3]
```diff
@@ -269,23 +284,45 @@ public CompositeContextNode subProcessNode(long id, String name, NodeContainer n
...
         return split;
     }
 
+    public Join joinNode(long id, String name, int type, NodeContainer nodeContainer) {
+        // 0 = TYPE_UNDEFINED
+        // 1 = TYPE_AND
+        // 2 = TYPE_XOR
+        // 3 = TYPE_DISCRIMINATOR
+        // 4 = TYPE_N_OF_M
+        // 5 = TYPE_OR
```
Comment (MarianMacik) [inline jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java:315]: Same here.

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java [hunk 4]
```diff
@@ -89,14 +92,39 @@ public static boolean includesSupportedStates(Workflow workflow) {
...
                     && !state.getType().equals(DefaultState.Type.RELAY)
-                    && !state.getType().equals(DefaultState.Type.SWITCH)) {
+                    && !state.getType().equals(DefaultState.Type.SWITCH)
+                    && !state.getType().equals(DefaultState.Type.PARALLEL)) {
                 return false;
             }
+
+            if(state.getType().equals(DefaultState.Type.PARALLEL)) {
+                if(!supportedParallelState((ParallelState) state)) {
+                    LOGGER.warn("unsupported parallel state");
```
Comment (MarianMacik) [inline jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java:102]: Can you please format this better and probably include a reason why it is not supported, i.e. just a subset of parallel states is supported etc.?

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/ServerlessWorkflowParser.java [hunk 5]
```diff
@@ -199,28 +216,64 @@ public Process parseWorkFlow(Reader workflowFile) throws JsonProcessingException
...
+                for(Branch branch : parallelState.getBranches()) {
+                    SubflowState subflowState = (SubflowState) branch.getStates().get(0);
+                    SubProcessNode callActivityNode = factory.callActivity(idCounter.getAndIncrement(), subflowState.getName(), subflowState.getWorkflowId(), subflowState.isWaitForCompletion(), process);
+
+                    factory.connect(parallelSplit.getId(), callActivityNode.getId(), parallelSplit.getId() + "_" + callActivityNode.getId(), process);
+                    factory.connect(callActivityNode.getId(), parallelJoin.getId(), callActivityNode.getId() + "_" + parallelJoin.getId(), process);
+
+                }
+
+                if(state.getStart() != null) {
```
[the remark this answers is not in the document]
Comment (tsurdilo) [reply jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/ServerlessWorkflowParser.java]: fixed - thanks!

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java (hunk 1 above)
Comment (tsurdilo) [reply jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java]: good point - fixed

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java (hunk 2 above)
Comment (tsurdilo) [reply jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java]: removed. thanks!

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java (hunk 3 above)
Comment (tsurdilo) [reply jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/core/ServerlessWorkflowFactory.java]: removed - thanks!

File: jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java (hunk 4 above)
Comment (tsurdilo) [reply jbpm/jbpm-serverless-workflow/src/main/java/org/jbpm/serverless/workflow/parser/util/ServerlessWorkflowUtils.java]: formatted and updated logger warn message. thanks

**Comment by tsurdilo** [comment]
@MarianMacik build passed fine -  let me know if there is anything else that needs to be updated.

**Comment by cristianonicolai** [comment]
@tsurdilo please just update master with the recent changes here.

**Pull request merged**